### ISSUE TITLE
NET-1868 S2190 FP: Yielding an infinite-length sequence

### ISSUE THREAD

**Issue by alexthornton1** [SonarSource/sonar-dotnet#5261]
### Description

When writing a method that uses `yield return` to (intentionally) yield an infinite-length sequence, SonarSource reported this warning:

```
S2190: Add a way to break out of this method's recursion.
```

### Repro steps

The following method is just about the simplest one that triggers the S2190 warning:

```csharp
public static IEnumerable<T> Repeat<T>(T element)
{
    while (true)
    {
        yield return element;
    }
}
```

### Expected behavior

If this same method had used `return` instead of `yield return`, then I would have expected to see S2190.  However, if `yield return` is used, then there is a way out of the infinite loop (by the code consuming the sequence choosing not to continue consuming it).  Calling this method is not necessarily a problem; materializing the sequence is.

I can certainly appreciate that this is a scenario that can reasonably be judged either way, so it's possible that the design choice here was an intentional one, in which case, of course, I'll understand and continue using the workaround.

### Actual behavior

The warning S2190 is produced.

### Known workarounds

The warning can be disabled in the source code file where a method intentionally generates an infinite-length sequence.

### Related information

* Visual Studio Enterprise 2022, version 17.0.4
* SonarLint for Visual Studio 2022 extension, version 5.3.0.41207
* .NET 6 / C# 10
* Windows 10

**Comment by Corniel** [comment]
I can confirm this issue. And I agree is should not report on this issue.

**Comment by pavel-mikula-sonarsource** [comment]
Hi,
While I understand the purpose of this very specific use case, I also think that it's exactly what the rule should also detect and highlight. As the consumer is not enforced to break out of the iteration.

You've mentioned that the reproducer is the simplest one. I assume you might have a more complicated scenario in your code. Can you try to propose a precise condition when we should try to not report on this case?

For now, I'm in favor of highlighting the scenario here as well and using one of the available suppression mechanisms (like accepting the issue as `Won't fix` with explanation comment on SonarCloud/SonarQube or using `#pragma warning disable`) to highlight and document the decision.

**Comment by alexthornton1** [comment]
I think it boils down to the amount of sympathy one has for the kinds of design techniques you'd see in lazy-evaluated functional languages like Haskell, where it's quite natural to write a function that intentionally generates infinite outputs, since the output isn't _actually_ infinite until a caller tries to consume it infinitely.  Haskell takes that idea to its logical conclusion, with basically _everything_ being lazy-evaluated, whereas C# applies the idea more surgically, with `IEnumerable<T>` being one of the key places in its design that provides it.  If a C# method returning `IEnumerable<T>` produces essentially a lazy-evaluated sequence of results, then it can produce an infinite lazy-evaluated sequence, which won't become infinite unless it's materialized (e.g., by calling `ToList` on it).

Some would reasonably argue that such a technique is not appropriate for C#, in which case one would also argue that SonarQube would be right to warn about it.  Others would reasonably argue that the technique is just as useful in C# as it is in Haskell, in which case it would be better for SonarQube _not_ to warn about it.  (This is why, when I reported the issue, I expected that the response to it could go either way; I can see a reasonable argument both ways.)

The use case where this warning arose for me wasn't much more complex than the one I presented in this issue, in fact.

- I have a (closed-source) library that implements retrying logic (i.e., keep trying until success or until we run out of retries, with delays between those retries) by accepting an `IEnumerable<TimeSpan>` that specifies the length of each delay.  When there are no more delays in that sequence, no more retries will be attempted.
- I needed the ability to retry an operation indefinitely, once per minute until success (or until the process was shut down entirely, which is handled separately via a `CancellationToken`).

The most natural expression of that idea was an infinite sequence of one-minute `TimeSpan`s, so I wrote this:

```csharp
public class InfiniteEnumerable
{
    public static IEnumerable<T> Repeat<T>(T element)
    {
        while (true)
        {
            yield return element;
        }
    }
}
```

And then passed `InfiniteEnumerable.Repeat(TimeSpan.FromSeconds(1))` as my retry sequence.

That approach worked a treat, except for the SonarQube warning, which I've disabled in that source file for now.

But given the general nature of the problem &mdash; I can't be the only person using an infinite-length sequence for good rather than evil :grinning: &mdash; I thought I'd run it by your team, in case it was something you'd want to change.  But, again, I'm totally sympathetic to the idea that this might be a situation that's better warned about than not.

**Comment by pavel-mikula-sonarsource** [comment]
Thanks @alexthornton1 for the explanation. As there was your case plus two others on the thread, I confirm it as a False Positive.

**Comment by pavel-mikula-sonarsource** [comment]
As late as now I've realized that the rule complains about recursion, but the code contains an infinite loop instead. Not a recursion.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1868](https://sonarsource.atlassian.net/browse/NET-1868)

[NET-1868]: https://sonarsource.atlassian.net/browse/NET-1868?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
